File: 542-01-matrix/01-matrix.java
class Pair{
    int i,j;
    Pair(int i,int j){
        this.i=i;
        this.j=j;
    }
}
class Solution {
    public int[][] updateMatrix(int[][] mat) {
        int m=mat.length,n=mat[0].length;
        int[][]vis=new int[m][n];
        Queue<Pair>q=new LinkedList<>();
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(mat[i][j]==0){
                    q.add(new Pair(i,j));
                }
                else{
                    vis[i][j]=Integer.MAX_VALUE;
                }
            }
        }
        while(q.size()>0){
            Pair p=q.remove();
            int i=p.i,j=p.j;
            if(i+1<m && mat[i+1][j]==1 && vis[i][j]+1<vis[i+1][j]){
                vis[i+1][j]=vis[i][j]+1;
                q.add(new Pair(i+1,j));
            }
            if(j+1<n && mat[i][j+1]==1 && vis[i][j]+1<vis[i][j+1]){
                vis[i][j+1]=vis[i][j]+1;
                q.add(new Pair(i,j+1));
            }
            if(i-1>=0 && mat[i-1][j]==1 && vis[i][j]+1<vis[i-1][j]){
                vis[i-1][j]=vis[i][j]+1;
                q.add(new Pair(i-1,j));
            }
            if(j-1>=0 && mat[i][j-1]==1 && vis[i][j]+1<vis[i][j-1]){
                vis[i][j-1]=vis[i][j]+1;
                q.add(new Pair(i,j-1));
            }

        }
        return vis;
    }
}

File: 542-01-matrix/01-matrix.cpp
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        int n=mat.size();
        int m=mat[0].size();
        vector<vector<bool>> isVisited(n,vector<bool>(m,false));
        queue<vector<int>> q;
        vector<vector<int>> temp={{1,0},{0,1},{-1,0},{0,-1}};
        for(int i=0;i<mat.size();i++){
            for(int j=0;j<mat[0].size();j++){
                if(mat[i][j]==0){
                    isVisited[i][j]=true;
                    q.push({i,j,0});
                }
            }
        }
        while(q.size()){
            vector<int> v=q.front();
            q.pop();
            int x=v[0];
            int y=v[1];
            int count=v[2];
            for(int i=0;i<temp.size();i++){
                int x_=x+temp[i][0];
                int y_=y+temp[i][1];
                if(x_>=0 && y_>=0 && x_<n && y_<m &&isVisited[x_][y_]==false){
                    mat[x_][y_]=count+1;
                    isVisited[x_][y_]=true;
                    q.push({x_,y_,count+1});
                }
            }
        }
        return mat;
    }
};